backend/comballoc.ml

```diff
@@ -30,16 +31,19 @@ let rec combine i allocstate =
   match i.desc with
     Iend | Ireturn _ | Iexit _ | Iraise _ ->
       (i, allocstate)
-  | Iop(Ialloc { bytes = sz; dbginfo; _ }) ->
+  | Iop(Ialloc { bytes = sz; dbginfo; mode }) ->
       assert (List.length dbginfo = 1);
       begin match allocstate with
-      | Pending_alloc {reg; dbginfos; totalsz}
-          when totalsz + sz <= (Config.max_young_wosize + 1) * Arch.size_addr ->
+      | Pending_alloc {reg; dbginfos; totalsz; mode = prev_mode}
+          when (mode = prev_mode) &&
+              ((totalsz + sz <= (Config.max_young_wosize + 1) * Arch.size_addr)
+               || mode = Lambda.Alloc_local) ->
```

Comment:
The result might be not that good if this alternates between local and non local allocs.
Wouldn't it be better to split the state into local and normal alloc ?

Reply:
I also thought so! That's what it used to do and it hit a bug.

The half-initialised allocation midway through a Comballoc block is not a valid state that the GC can see. So if you do a GC alloc in the middle of a Comballoc'd sequence of local allocs, the GC can enter and crash on the malformed local allocation. See https://github.com/janestreet/ocaml/pull/41